tests/MongoDB.Driver.Tests/Specifications/client-side-encryption/ClientSideEncryptionTestRunner.cs

```diff
@@ -0,0 +1,316 @@
+                            kmsOptions.Add("accessKeyId", awsAccessKey);
+                            kmsOptions.Add("secretAccessKey", awsSecretAccessKey);
+                        }
+                        providers.Add(kmsProvider.Name, kmsOptions);
+                        break;
+                    case "local":
+                        if (kmsProvider.Value.AsBsonDocument.TryGetElement("key", out var key))
+                        {
+                            var binary = key.Value.AsBsonBinaryData;
+                            kmsOptions.Add(key.Name, binary.Bytes);
+                        }
+                        providers.Add(kmsProvider.Name, kmsOptions);
+                        break;
+                    default:
+                        throw new Exception($"Unexpected kms provider type {kmsProvider.Name}.");
+                }
+            }
+
+            return new ReadOnlyDictionary<string, IReadOnlyDictionary<string, object>>(providers);
+        }
+
+        public void ReplaceTypeAssertionWithActual(BsonDocument actual, BsonDocument expected)
+        {
+            for (int i = 0; i < expected.ElementCount; i++)
+            {
+                var expectedElement = expected.ElementAt(i);
+                var value = expectedElement.Value;
+                if (value.IsBsonDocument)
+                {
+                    var valueDocument = value.AsBsonDocument;
+                    var actualValue = actual.GetValue(expectedElement.Name, null);
+                    if (valueDocument.ElementCount == 1 && valueDocument.Select(c => c.Name).Single().Equals("$$type"))
+                    {
+                        var type = valueDocument["$$type"].AsString;
+                        if (type.Equals("binData"))
+                        {
+                            expected[expectedElement.Name] = actualValue;
+                        }
+                        else if (type.Equals("long"))
+                        {
+                            expected[expectedElement.Name] = actualValue;
+                        }
+                    }
+                    else if (actualValue != null && actualValue.IsBsonDocument)
+                    {
+                        ReplaceTypeAssertionWithActual(actualValue.AsBsonDocument, valueDocument);
+                    }
+                }
+                else if (value.IsBsonArray)
+                {
+                    ReplaceTypeAssertionWithActual(actual[expectedElement.Name].AsBsonArray, value.AsBsonArray);
+                }
+            }
+        }
+
+        private void ReplaceTypeAssertionWithActual(BsonArray actual, BsonArray expected)
+        {
+            for (int i = 0; i < expected.Count; i++)
+            {
+                var value = expected.ElementAt(i);
+                if (value.IsBsonDocument)
+                {
+                    ReplaceTypeAssertionWithActual(actual.ElementAt(i).AsBsonDocument, value.AsBsonDocument);
```

Comment:
Why `ElementAt(i)`? You can use this:

    actual[i]

which avoids the overhead of the LINQ `ElementAt` method.

Reply:
done